tt_metal/api/program.hpp

```diff
@@ -0,0 +1,128 @@
+
+
+/**
+ * @brief Initializes a semaphore on specified cores.
+ *
+ * @param program The program to which the semaphore will be added.
+ * @param core_spec Range of cores using the semaphore.
+ * @param initial_value Initial value of the semaphore.
+ * @param core_type Core type on which to create the semaphore (default: CoreType::WORKER).
+ * @return Semaphore address as a uint32_t.
+ */
+uint32_t CreateSemaphore(
+    Program program,
+    const CoreRangeSet &core_spec,
+    uint32_t initial_value,
+    CoreType core_type = CoreType::WORKER);
+
+
+/**
+ * @brief Creates a Circular Buffer in L1 memory of specified cores and adds it to the program.
+ *
+ * @param program The program to which the buffer will be added.
+ * @param core_spec Specifies the cores where the circular buffer will be configured.
+ * @param config Configuration for the circular buffer.
+ * @return CBHandle representing the Circular Buffer ID.
+ */
+CBHandle CreateCircularBuffer(
+    Program program,
+    const CoreRangeSet &core_spec,
+    const CircularBufferConfig &config);
+
+/**
+ * @brief Gets the configuration of a circular buffer.
+ *
+ * @param program The program containing the circular buffer.
+ * @param cb_handle Handle of the circular buffer.
+ * @return Reference to the CircularBufferConfig.
+ */
+const CircularBufferConfig &GetCircularBufferConfig(Program program, CBHandle cb_handle);
+
+/**
+ * @brief Updates the total size of a circular buffer.
+ *
+ * @param program The program containing the circular buffer.
+ * @param cb_handle Handle of the circular buffer.
+ * @param total_size New total size of the circular buffer in bytes.
+ */
+void UpdateCircularBufferTotalSize(Program program, CBHandle cb_handle, uint32_t total_size);
+
+/**
+ * @brief Updates the address of a dynamic circular buffer.
+ *
+ * @param program The program containing the circular buffer.
+ * @param cb_handle Handle of the circular buffer.
+ * @param buffer Dynamically allocated L1 buffer that shares address space with the circular buffer.
+ */
+void UpdateDynamicCircularBufferAddress(Program program, CBHandle cb_handle, const Buffer buffer);
+
+
+//==================================================
+//                 PROGRAM FUNCTIONS
+//==================================================
+
+/**
+ * @brief Retrieves the circular buffers associated with the program.
+ *
+ * @param program The program to query.
+ * @return Reference to a vector of shared pointers to CircularBuffer objects.
+ */
+const std::vector<std::shared_ptr<CircularBuffer>> &CircularBuffers(Program program);
+
+/**
+ * @brief Retrieves the circular buffers associated with the program on a specific core range.
+ *
+ * @param program The program to query.
+ * @param cr The core range to consider.
+ * @return Vector of shared pointers to CircularBuffer objects on the core range.
+ */
+const std::vector<std::shared_ptr<CircularBuffer>> CircularBuffersOnCorerange(Program program, const CoreRange &cr);
```

Comment:
`const`-qualifying the return type seems to me that it would impose unnecessary restrictions on usage, such as requiring a copy construction if you want to modify it. Either `std::vector<CircularBuffer>` or `const std::vector<CircularBuffer> &` would make more sense, and if the latter is chosen, I would even lean towards `std::span<const CircularBuffer>`.